internal/node/spawn_procman.go

```diff
@@ -33,50 +64,215 @@ func NewSpawningProcessManager(
+			}
+
+			s.liveProcs[p.ID] = p
+			s.stopMutexes[p.ID] = &sync.Mutex{}
+
+			go s.delegate.OnProcessStarted(p.ID)
 
-	// return nil
-	panic("Not implemented")
+			s.log.Info("Adding new agent process to warm pool",
+				slog.String("workload_id", p.ID))
+
+			s.warmProcs <- p // If the pool is full, this line will block until a slot is available.
+		}
+	}
+
+	return nil
 }
 
+// Stops a single agent process
 func (s *SpawningProcessManager) StopProcess(workloadID string) error {
-	// TODO
+	proc, exists := s.liveProcs[workloadID]
+	if !exists {
+		return fmt.Errorf("failed to stop process %s. No such process", workloadID)
+	}
+
+	delete(s.deployRequests, workloadID)
+
+	mutex := s.stopMutexes[workloadID]
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	s.log.Debug("Attempting to stop process", slog.String("workload_id", workloadID))
 
-	panic("Not implemented")
+	err := s.kill(proc)
+	if err != nil {
+		return err
+	}
+
+	delete(s.liveProcs, workloadID)
+	delete(s.stopMutexes, workloadID)
+
+	return nil
 }
 
+// Looks up an agent process. A non-existent agent process returns (nil, nil), not
+// an error
 func (s *SpawningProcessManager) Lookup(workloadID string) (*agentapi.DeployRequest, error) {
-	// TODO
+	if request, ok := s.deployRequests[workloadID]; ok {
+		return request, nil
+	}
+
+	// Per contract, a non-prepared workload returns nil, not error
+	return nil, nil
+}
+
+// Checks if the process manager is stopping
+func (s *SpawningProcessManager) stopping() bool {
+	return (atomic.LoadUint32(&s.closing) > 0)
+}
+
+// Spawns a new child process, a waiting nex-agent
+func (s *SpawningProcessManager) spawn() (*spawnedProcess, error) {
+	id := xid.New()
+	workloadID := id.String()
+
+	cmd := exec.Command(nexAgentBinary)
+	cmd.Env = append(os.Environ(),
+		"NEX_SANDBOX=false",
+		fmt.Sprintf("NEX_WORKLOADID=%s", workloadID),
+		// can't use the CNI host because we don't use it in no-sandbox mode
+		"NEX_NODE_NATS_HOST=0.0.0.0",
+		fmt.Sprintf("NEX_NODE_NATS_PORT=%d", *s.config.InternalNodePort),
+	)
+
+	cmd.Stderr = &procLogEmitter{workloadID: workloadID, log: s.log, stderr: true}
+	cmd.Stdout = &procLogEmitter{workloadID: workloadID, log: s.log, stderr: false}
+
+	newProc := &spawnedProcess{
+		ID:   workloadID,
+		cmd:  cmd,
+		log:  s.log,
+		Fail: make(chan bool),
+		Run:  make(chan bool),
+		Exit: make(chan int),
+	}
+
+	go func() {
+		err := cmd.Run()
+		if err != nil {
+			s.log.Info("Command finished with error", slog.Any("error", err))
+		}
+	}()
+
+	return newProc, nil
+}
+
+func (s *SpawningProcessManager) kill(proc *spawnedProcess) error {
+	if proc.cmd.Process != nil {
+		err := proc.cmd.Process.Signal(os.Kill)
+		if err != nil {
+			s.log.Error("Failed to kill OS process",
+				slog.String("workload_id", proc.ID),
+				slog.Int("pid", proc.cmd.Process.Pid))
+			return err
+		}
+	}
+
+	return nil
+}
+
+type procLogEmitter struct {
+	stderr bool
+	// TODO: personal opinion - not sure I like propagating logger instances everywhere...
```

Comment:
Agree, I don't think it would be the end of the world to reference a global `Log` instance that was setup on package `init`.

Reply:
oof, its not often I am on the opposite side of the line!  I would really not use globals at all if possible.  You instantly loose context and don't gain anything in return

Reply:
Logging is about the only thing that I will compromise on. I don't mind passing it around though, for the reasons you stated 😏